File: tools/redownload_plan_images.js
/**
 * Re-download unique per-plan images from Century Communities plan pages.
 * Overwrites cover.jpg / 01.jpg / 02.jpg / 03.jpg under uploads/properties/{slug}/
 */
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { URL } = require('url');

const ROOT = path.resolve(__dirname, '..');
const INV = path.join(__dirname, 'parent_company_inventory.json');
const OUT_MAP = path.join(__dirname, 'plan_image_urls.json');
const LOG = path.join(__dirname, 'plan_image_redownload.log');

const UA =
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36';

const ASSET_RE =
  /https?:\/\/[^"'\\\s<>]+globalassets[^"'\\\s<>]+\.(?:jpe?g|png|webp)/gi;
const REL_ASSET_RE = /\/globalassets\/[^"'\\\s<>]+\.(?:jpe?g|png|webp)/gi;

function normalize(u) {
  return u.replace(/\\u0026/g, '&').replace(/&amp;/g, '&').split('?')[0];
}

async function fetchText(url) {
  const res = await fetch(url, {
    headers: {
      'User-Agent': UA,
      Accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
      'Accept-Language': 'en-US,en;q=0.9',
    },
  });
  if (!res.ok) throw new Error(`HTTP ${res.status} for ${url}`);
  return await res.text();
}

async function fetchBin(url) {
  const res = await fetch(url, { headers: { 'User-Agent': UA } });
  if (!res.ok) throw new Error(`HTTP ${res.status} for ${url}`);
  const buf = Buffer.from(await res.arrayBuffer());
  return buf;
}

function extractAssets(html) {
  const found = new Set();
  for (const m of html.matchAll(ASSET_RE)) found.add(normalize(m[0]));
  for (const m of html.matchAll(REL_ASSET_RE)) {
    found.add(normalize('https://www.centurycommunities.com' + m[0]));
  }
  return [...found];
}

function scoreUrl(url, planSlug, planName, forCover = false) {
  const u = url.toLowerCase();
  const slug = String(planSlug || '').toLowerCase();
  const name = String(planName || '')
    .replace(/^the\s+/i, '')
    .toLowerCase()
    .replace(/\s+/g, '-');
  if (/icon-|interface-|footer|eliant|svg$|logo/.test(u)) return -1000;
  let score = 0;
  const isElev = /elev|elevation|exterior|extlike|_ext|ext_/.test(u);
  const isInterior = /interior|furnished|kitchen|bedroom|great-room|primary|dining/.test(u);
  const isFloor = /floor-\d|floor_1|floor_2|floorplan|floor-plan/.test(u) && !isElev;
  if (isElev) score += forCover ? 120 : 70;
  if (isInterior) score += forCover ? 40 : 90;
  if (isFloor) score += forCover ? -40 : 15;
  if (slug && u.includes(slug)) score += 80;
  if (name && u.includes(name)) score += 60;
  if (u.includes('aerial')) score += forCover ? 30 : 10;
  if (u.includes('model-photos')) score += forCover ? 50 : 30;
  return score;
}

function pickImages(urls, planSlug, planName, need = 4) {
  const rank = (forCover) =>
    urls
      .map((u) => ({ u, s: scoreUrl(u, planSlug, planName, forCover) }))
      .filter((x) => x.s > 0)
      .sort((a, b) => b.s - a.s);

  const coverRanked = rank(true);
  const galleryRanked = rank(false);
  const picked = [];
  const seen = new Set();

  const take = (list) => {
    for (const { u } of list) {
      const base = path
        .basename(new URL(u).pathname)
        .replace(/\.[^.]+$/, '')
        .toLowerCase();
      if (seen.has(base)) continue;
      seen.add(base);
      picked.push(u);
      if (picked.length >= need) return true;
    }
    return false;
  };

  // Cover first from elevation-biased ranking, then fill gallery.
  take(coverRanked.slice(0, 1));
  take(galleryRanked);
  if (picked.length < need) take(coverRanked);
  return picked;
}

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

async function main() {
  const inv = JSON.parse(fs.readFileSync(INV, 'utf8'));
  const results = [];
  const lines = [`Started ${new Date().toISOString()}`];

  for (const p of inv.properties) {
    const { slug, plan_slug: planSlug, plan, url } = p;
    console.log(`=== ${slug} (${plan}) ===`);
    let html;
    try {
      html = await fetchText(url);
    } catch (e) {
      const msg = `FAIL fetch ${slug}: ${e.message}`;
      console.warn(msg);
      lines.push(msg);
      results.push({ slug, ok: false, error: e.message });
      continue;
    }
    const assets = extractAssets(html);
    console.log(`  assets found: ${assets.length}`);
    const picked = pickImages(assets, planSlug, plan, 4);
    if (!picked.length) {
      const msg = `FAIL: no plan images for ${slug}`;
      console.warn(msg);
      lines.push(msg);
      results.push({ slug, ok: false, error: 'no images', assets: assets.length });
      continue;
    }

    const destDir = path.join(ROOT, 'uploads', 'properties', slug);
    fs.mkdirSync(destDir, { recursive: true });
    const names = ['cover.jpg', '01.jpg', '02.jpg', '03.jpg'];
    const saved = [];
    for (let i = 0; i < names.length; i++) {
      const src = picked[i] || picked[0];
      const dest = path.join(destDir, names[i]);
      try {
        const buf = await fetchBin(src);
        fs.writeFileSync(dest, buf);
        console.log(`  ${names[i]} <- ${src} (${buf.length} bytes)`);
        saved.push({ file: names[i], url: src, bytes: buf.length });
        lines.push(`${slug} ${names[i]} ${src}`);
      } catch (e) {
        const msg = `FAIL download ${slug} ${names[i]}: ${e.message}`;
        console.warn(msg);
        lines.push(msg);
      }
    }
    results.push({ slug, plan, url, ok: saved.length > 0, image_urls: picked, saved });
    await sleep(350);
  }

  const groups = new Map();
  for (const dir of fs.readdirSync(path.join(ROOT, 'uploads', 'properties'))) {
    if (!dir.startsWith('cc-')) continue;
    const cover = path.join(ROOT, 'uploads', 'properties', dir, 'cover.jpg');
    if (!fs.existsSync(cover)) continue;
    const h = crypto.createHash('md5').update(fs.readFileSync(cover)).digest('hex');
    if (!groups.has(h)) groups.set(h, []);
    groups.get(h).push(dir);
  }

  console.log('--- cover hash groups ---');
  const dupes = [];
  let unique = 0;
  for (const [h, slugs] of [...groups.entries()].sort((a, b) => b[1].length - a[1].length)) {
    const line = `${slugs.length}x ${h} -> ${slugs.join(', ')}`;
    console.log(line);
    lines.push(line);
    if (slugs.length === 1) unique += 1;
    else dupes.push({ hash: h, count: slugs.length, slugs });
  }

  const payload = {
    generated: new Date().toISOString(),
    unique_cover_hashes: unique,
    duplicate_cover_groups: dupes,
    properties: results,
  };
  fs.writeFileSync(OUT_MAP, JSON.stringify(payload, null, 2));
  fs.writeFileSync(LOG, lines.join('\n') + '\n');
  console.log(`Wrote ${OUT_MAP}`);
  console.log(`Unique covers: ${unique} / ${results.length}`);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: tools/redownload_plan_images.py
#!/usr/bin/env python3
"""Re-download unique per-plan images from Century Communities plan pages."""

from __future__ import annotations

import hashlib
import json
import re
import time
import urllib.error
import urllib.request
from collections import defaultdict
from pathlib import Path
from urllib.parse import urlparse

ROOT = Path(__file__).resolve().parents[1]
INV = Path(__file__).with_name("parent_company_inventory.json")
OUT_MAP = Path(__file__).with_name("plan_image_urls.json")
LOG = Path(__file__).with_name("plan_image_redownload.log")

UA = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36"
)

ASSET_RE = re.compile(
    r"""https?://[^"'\\\s<>]+globalassets[^"'\\\s<>]+\.(?:jpe?g|png|webp)""",
    re.I,
)
REL_ASSET_RE = re.compile(
    r"""/globalassets/[^"'\\\s<>]+\.(?:jpe?g|png|webp)""",
    re.I,
)


def fetch_html(url: str) -> str:
    req = urllib.request.Request(
        url,
        headers={
            "User-Agent": UA,
            "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
            "Accept-Language": "en-US,en;q=0.9",
        },
    )
    with urllib.request.urlopen(req, timeout=60) as resp:
        return resp.read().decode("utf-8", errors="replace")


def normalize(url: str) -> str:
    url = url.replace("\\u0026", "&").replace("&amp;", "&")
    return url.split("?", 1)[0]


def extract_assets(html: str) -> list[str]:
    found: set[str] = set()
    for m in ASSET_RE.finditer(html):
        found.add(normalize(m.group(0)))
    for m in REL_ASSET_RE.finditer(html):
        found.add(normalize("https://www.centurycommunities.com" + m.group(0)))
    return sorted(found)


def score_url(url: str, plan_slug: str, plan_name: str) -> int:
    u = url.lower()
    slug = plan_slug.lower()
    name = re.sub(r"^the\s+", "", plan_name, flags=re.I).lower().replace(" ", "-")
    if re.search(r"icon-|interface-|footer|eliant|svg$|logo", u):
        return -1000
    score = 0
    if re.search(r"elev|elevation|exterior|ext_|model-photos|_ext", u):
        score += 50
    if re.search(r"interior|furnished|kitchen|bedroom|great-room|primary", u):
        score += 20
    if slug and slug in u:
        score += 80
    if name and name in u:
        score += 60
    if re.search(r"floor-\d|floorplan|floor_plan", u):
        score += 5
    if "aerial" in u:
        score += 10
    return score


def pick_images(urls: list[str], plan_slug: str, plan_name: str, need: int = 4) -> list[str]:
    ranked = []
    for u in urls:
        s = score_url(u, plan_slug, plan_name)
        if s > 0:
            ranked.append((s, u))
    ranked.sort(key=lambda x: x[0], reverse=True)
    picked: list[str] = []
    seen: set[str] = set()
    for _, u in ranked:
        base = Path(urlparse(u).path).stem.lower()
        if base in seen:
            continue
        seen.add(base)
        picked.append(u)
        if len(picked) >= need:
            break
    return picked


def download(url: str, dest: Path) -> int:
    dest.parent.mkdir(parents=True, exist_ok=True)
    tmp = dest.with_suffix(dest.suffix + ".part")
    req = urllib.request.Request(url, headers={"User-Agent": UA})
    with urllib.request.urlopen(req, timeout=90) as resp, open(tmp, "wb") as f:
        data = resp.read()
        f.write(data)
    tmp.replace(dest)
    return len(data)


def main() -> None:
    inv = json.loads(INV.read_text(encoding="utf-8"))
    results = []
    lines: list[str] = [f"Started"]

    for p in inv["properties"]:
        slug = p["slug"]
        plan_slug = p["plan_slug"]
        plan = p["plan"]
        url = p["url"]
        print(f"=== {slug} ({plan}) ===")
        try:
            html = fetch_html(url)
        except Exception as e:
            msg = f"FAIL fetch {slug}: {e}"
            print(msg)
            lines.append(msg)
            results.append({"slug": slug, "ok": False, "error": str(e)})
            continue

        assets = extract_assets(html)
        print(f"  assets found: {len(assets)}")
        picked = pick_images(assets, plan_slug, plan, 4)
        if not picked:
            msg = f"FAIL: no plan images for {slug}"
            print(msg)
            lines.append(msg)
            results.append({"slug": slug, "ok": False, "error": "no images", "assets": len(assets)})
            continue

        dest_dir = ROOT / "uploads" / "properties" / slug
        names = ["cover.jpg", "01.jpg", "02.jpg", "03.jpg"]
        saved = []
        for i, name in enumerate(names):
            src = picked[i] if i < len(picked) else picked[0]
            dest = dest_dir / name
            try:
                nbytes = download(src, dest)
                print(f"  {name} <- {src} ({nbytes} bytes)")
                saved.append({"file": name, "url": src, "bytes": nbytes})
                lines.append(f"{slug} {name} {src}")
            except Exception as e:
                msg = f"FAIL download {slug} {name}: {e}"
                print(msg)
                lines.append(msg)
        results.append(
            {
                "slug": slug,
                "plan": plan,
                "url": url,
                "ok": bool(saved),
                "image_urls": picked,
                "saved": saved,
            }
        )
        time.sleep(0.35)

    # uniqueness
    groups: dict[str, list[str]] = defaultdict(list)
    for cover in sorted((ROOT / "uploads" / "properties").glob("cc-*/cover.jpg")):
        h = hashlib.md5(cover.read_bytes()).hexdigest()
        groups[h].append(cover.parent.name)

    dupes = {h: slugs for h, slugs in groups.items() if len(slugs) > 1}
    unique = sum(1 for slugs in groups.values() if len(slugs) == 1)
    print("--- cover hash groups ---")
    for h, slugs in sorted(groups.items(), key=lambda kv: (-len(kv[1]), kv[0])):
        line = f"{len(slugs)}x {h} -> {', '.join(slugs)}"
        print(line)
        lines.append(line)

    payload = {
        "unique_cover_hashes": unique,
        "duplicate_cover_groups": [
            {"hash": h, "count": len(s), "slugs": s} for h, s in dupes.items()
        ],
        "properties": results,
    }
    OUT_MAP.write_text(json.dumps(payload, indent=2), encoding="utf-8")
    LOG.write_text("\n".join(lines) + "\n", encoding="utf-8")
    print(f"Wrote {OUT_MAP}")
    print(f"Unique covers: {unique} / {len(results)}")


if __name__ == "__main__":
    main()
